imagebuildah/executor.go

```diff
@@ -352,6 +357,67 @@ func (b *Executor) resolveNameToImageRef(output string) (types.ImageReference, e
 	return imageRef, err
 }
 
+// buildCustomOutput takes imageID of the build and exports final built rootfs to specified directory or tar file.
+func (b *Executor) buildCustomOutput(ctx context.Context, imageID string, isDir bool, isStdout bool, path string) error {
+	didMount := false
+	image, err := internalUtil.LookupImage(b.systemContext, b.store, imageID)
+	if err != nil {
+		return errors.Wrapf(err, "failed while looking image %q", imageID)
+	}
+	defer func() {
+		if didMount {
+			// Log error as debug log here, since image could have been unmounted already
+			// or image is being used by another build, there is nothing we can do.
+			err := image.Unmount(false)
+			if err != nil {
+				logrus.Debugf("Unmount failed with: %v", err)
+			}
+		}
+	}()
+	mountPoint, err := image.Mountpoint()
+	if err != nil {
+		return errors.Wrapf(err, "failed getting mount point of image %q", imageID)
+	}
+	if mountPoint == "" {
+		didMount = true
+		mountPoint, err = image.Mount(ctx, nil, "")
+		if err != nil {
+			return errors.Wrapf(err, "failed mounting image %q", imageID)
+		}
+	}
+	input, err := archive.Tar(mountPoint, archive.Uncompressed)
+	if err != nil {
+		return errors.Wrapf(err, "failed while performing tar for %q", mountPoint)
+	}
+	if isDir {
+		if !filepath.IsAbs(path) {
+			cwd, err := os.Getwd()
+			if err != nil {
+				return err
+			}
+			path = filepath.Join(cwd, filepath.Clean(path))
+		}
+		err = archive.Untar(input, path, nil)
```

Comment:
How are ownerships intended to work for unprivileged users, where this happens inside of a user namespace?

Reply:
I'm expecting ownership to stay as it is which was set in image and it should not be changed for unprivileged users so image out can become read-only if files where meant to be written for some other user. 

But I'll verify behavior once with rootless buildkit as well just to stay in parity.

Reply:
This would need to be verified.

Reply:
What does that mean?  When we write content to a directory, who ends up owning files that are and aren't owned by root in the image?  Are setuid/setgid bits and extended attributes preserved?

Reply:
So when i export content with buildkit ownership does not stays as is everything is changing to current user so for example if I do 

```Dockerfile
FROM alpine
COPY hello .
RUN chown 1000:1000 hello
```

Exported content here has ownership of `hello` changed to `root:root` on host fs which I expected not to happen and stayed as `1000:1000` so just to match the behviour I added `NoLchown: true`.

Reply:
I have a hunch that either buildkit is not doing something right here and we are just trying to match the parity or maybe it is done intentionally to match some practical use-case.

Reply:
@nalind Following behavior also states true for `unprivileged user` an easier test is to run `buildkit:master-rootless` image.
So a reproducer is something like
```console
docker run --name buildkitd -it --privileged moby/buildkit:master-rootless
docker exec -it buildkitd sh
whoami
# should be `user` and not `root`
cd ~
touch hello
```

Inside container create containerfile

```Dockerfile
FROM alpine
COPY hello .
RUN chown 1000:1000 hello
```
and run
```
buildctl build --frontend=dockerfile.v0  --local context=. --local dockerfile=. --output type=local,dest=output-dir`
```
all the files inside `output-dir` have been chowned to user `user:user`.


Currently buildah follows this exact same behavior.

Reply:
Are setuid/setgid bits and extended attributes preserved?

Reply:
@nalind Yes both are preserved in both buildkit and buildah verified manually with `setfattr` and `getfattr` after unpacking.

Reply:
Just adding a note here with a reproducer shared by @nalind , `buildkit` has inconsistent and not documented behavior between `privileged` and `unprivileged` invocation and permissions are produced differently for binaries like `/usr/bin/write` in both invocation.

So something like 
```Dockerfile
FROM registry.fedoraproject.org/fedora:35
RUN rpm --setcaps shadow-utils
RUN dnf -y install attr util-linux && dnf clean all
RUN chmod u+rwx `rpm -ql filesystem` 2> /dev/null || true
RUN echo test > /test.txt && chown 1000:1000 /test.txt
RUN getcap -v /usr/bin/newuidmap /usr/bin/newgidmap ; stat -c "%A   %n" /test.txt /usr/bin/su /usr/bin/write /usr/bin/sudo
``` 

With buildkit exported output
```
#root
-rwxr-sr-x usr/bin/write
# unprivileged user
-rwxr-xr-x usr/bin/write
```

With buildah we share permissions similarly between invocation by privileged and unprivileged user 
```
#root
-rwxr-sr-x   usr/bin/write
# unpriviledged user
-rwxr-sr-x   usr/bin/write
```

Here behavior of buildah is different from buildkit a little bit but I was unable to find a reason why should we alter permission for unprivileged user hence leaving a note here so in future if we ever face this we can revisit this discussion.

Reply:
Still not entirely sure about what should we do here and which behavior is the correct one.

Reply:
Looked at it again it seems like the issue is with `unpriviledged` invocation of buildkit it does not preserves `setuid` but preserves `setgid` when manually added in Dockerfile but again overrides `setgid` for the files which were already there in the base image. This behavior looks really confusing to me could it be a bug in buildkit ? 

`priviledged` invocation of both buildah and buildkit works in the same manner and `unpriviledged` invocation of buildah preserves everything.

@nalind can we hit any security issues if we choose not to replicate buildkit's unpriviledged invocation behavior ?

Reply:
Either there's some rule there that we don't understand, or it's a bug.  `sudo` seems to be an outlier, as I also lose the setuid bit on binaries like passwd (from the passwd package), chfn and chsh (from the util-linux-user package), and the setgid bit on binaries like locate (from the mlocate package) and lockdev (from the lockdev package) when I run a build as an unprivileged user.
The general case sure seems to suggest that we should be stripping the setuid/setgid/sticky bits when we're writing to "local" destinations as an unprivileged user.

Reply:
Also, pretty sure we want to be preserving ownership information when we're invoked as root, even though buildkit seems to change setgid-"tty" /usr/bin/write and setgid-"lock" /usr/sbin/lockdev into setgid-"root" /usr/bin/write and /usr/sbin/lockdev.

Reply:
@nalind Modified buildah to preserve ownership as is when invoked with as `root`, this is different behavior from buildkit but looks good to me.

Reply:
They are preserved as is.

Reply:
The buildkit version also seems to create the top-level destination directory with 0700 permissions if it doesn't already exist, while we seem to create it as 0755 (neither modifies the permissions on an already-present directory).  If we're going to be putting setuid/setgid files in there, I'd be more comfortable if we used 0700.

Reply:
nice catch, fixed that as well.

Reply:
I still think that what we're seeing with `sudo`'s setuid bit being preserved when buildkit is run by an unprivileged user is not what we want to be using as our rule for everything.  We should be stripping setuid/setgid bits and capability xattrs when we're writing to "local" destinations as a non-root user.

Reply:
Pushed a new version which covers this for `non-root` user.